Documents API, objects, and DB implementation

fmontei opened the pull request.

Commit:
Initial engine framework

This commit adds the initial engine framework for Deckhand. Included
is the logic for parsing YAML files as well as validating them and
doing forward substitution as specified by the YAML file.

This commit also includes unit tests for the framework changes.

Commit:
Add jsonschema validation to Deckhand

Previously Deckhand was using manual validation of YAML files.
However, schema validation is much cleaner, robust and thorough.
So, this commit adds jsonschema as a dependence in Deckhand.
Tempest already uses it as a dependence as well -- so a precedence
already exists.

This commit also updates unit tests as needed. Documentation
changes will be made in a follow-up patch.

Commit:
Add additional documentation.

Commit:
Use regexes for jsonschema pre-validation

This commit adds the ability to dynamically pick the right
schema to validate the YAML data against, based on the apiVersion
included in the YAML metadata.

This commit also corrects the pre-validation schema by changing
"choices" to "pattern", allowing Deckhand to simply check for a
regex rather than having to enumerate all possible types of data.

fmontei committed:
Trivial fix to default_schema

Added parenthesis that was missing to regex.

Commit:
Update default_schema with our updated schema definition

This commit updates the default_schema based on recent changes
to the document definition that Deckhand is supposed to validate.

Commit:
Remove apiVersion attribute from substitutions.src attributes

The src apiVersion attribute will be dynamically validated based
on the loaded schemas.

Commit:
Remove apiVersion attribute from substitutions.src attributes

The src apiVersion attribute will be dynamically validated based
on the loaded schemas.

Commit:
Change apiVersion references to schemaVersion

Commit:
Add kind param to SchemaVersion class.

Commit:
Deckhand initial ORM implementation

This commit adds the initial implementation for the Deckhand ORM.
It is fashioned after Drydock's ORM, which in turn follows
OpenStack conventions.

This commit specifically:
  - creates the deckhand.objects package
  - includes base ORM classes
  - includes document ORM class which consists of the following
    fields:
      - blob: the document payload itself stored as an object
              that adheres to the bare minimum schema of
              schemaVersion, kind, metadata and data.
      - name: the name of the document, nullable
      - revision_index: the revision number for the document
      - status: the status of the document
  - includes basic fields for documents, i.e. what status
    the document is in

Commit:
Deckhand initial ORM implementation

This commit adds the initial implementation for the Deckhand ORM.
It is fashioned after Drydock's ORM, which in turn follows
OpenStack conventions.

This commit specifically:
  - creates the deckhand.objects package
  - includes base ORM classes
  - includes document ORM class which consists of the following
    fields:
      - blob: the document payload itself stored as an object
              that adheres to the bare minimum schema of
              schemaVersion, kind, metadata and data.
      - name: the name of the document, nullable
      - revision_index: the revision number for the document
      - status: the status of the document
  - includes basic fields for documents, i.e. what status
    the document is in

Commit:
Fix incorrect comment.

Commit:
[WIP] Implement documents API

This commit adds the documents API and adds logic for performing
pre-validation schema checking wherever applicable in the
documents API.

The following endpoints in the documents API have been implemented:
  - POST /documents

Commit:
Added control (API) readme.

Commit:
Initial DB API models implementation.

Commit:
Temporary change - do not commit

Commit:
Update README

intlabs commented on requirements.txt:
```diff
@@ -1,13 +1,18 @@
 falcon==1.1.0
 
+SQLAlchemy!=1.1.5,!=1.1.6,!=1.1.7,!=1.1.8,>=1.0.10 # MIT
```
is this required? https://github.com/openstack/oslo.db/blob/4.24.1/requirements.txt#L11

fmontei replied:
No, this sort of pattern is not required. I'll have to clean up all this stuff in a future commit.

mark-burnett commented on deckhand/db/api_models.py:
```diff
@@ -0,0 +1,65 @@
+    revision_index = sa.Column(sa.Integer, nullable=False)
+    document_schema = sa.Column(sa.String(64), nullable=False)
+    instance_key = sa.Column(sa.String(64), nullable=False, unique=True)
+    document_metadata = sa.Column(JSONEncodedDict(), nullable=False)
```
I like this overall table definition.

I think we should probably use postgres's JSONB type directly though rather than encoding and decoding it in the server.  I don't think that we need to worry about supporting other databases, even for testing.

With docker, it's reasonably easy to run a small postgres instance even for unit tests.

mark-burnett commented on deckhand/db/api_models.py:
```diff
@@ -0,0 +1,65 @@
+class Document(API_BASE):
+    __tablename__ = 'document'
+
+    id = sa.Column(sa.String(255), primary_key=True, autoincrement=True)
```
I think this can just be an integer.

mark-burnett commented on README.rst:
```diff
@@ -8,6 +8,6 @@ To run::
 	$ virtualenv -p python3 /var/tmp/deckhand
 	$ . /var/tmp/deckhand/bin/activate
-	$ pip install .
+	$ sudo pip install .
```
I guess this sudo slipped back in.  Maybe we can merge some of your older PRs once they're out of merge conflict so that you can rebase onto a more current master (I assume that's what happened)?

mark-burnett commented on deckhand/control/documents.py:
```diff
@@ -0,0 +1,79 @@
+        document_data = req.stream.read(req.content_length or 0)
+
+        try:
+            document = yaml.safe_load(document_data)
```
I think you will want `yaml.safe_load_all`, since we will probably post many documents at once.

fmontei replied:
Agreed. Fixed in a future commit.

mark-burnett commented on deckhand/control/documents.py:
```diff
@@ -0,0 +1,79 @@
+
+    def on_post(self, req, resp):
+        """Create a document. Accepts YAML data only."""
+        if req.content_type != 'application/yaml':
```
`application/yaml` is not an official media type.  Other frameworks have used `application/x-yaml`, which I suggest we also follow.

fmontei replied:
Nice catch, I noticed you added that in your design document, I just haven't gotten around to changing it. I'll use a global variable so that way this sort of typo won't be made in the future.

mark-burnett commented on deckhand/db/api_models.py:
```diff
@@ -0,0 +1,65 @@
+        return value
+
+
+class Document(API_BASE):
```
It's fine to use the ORM if you prefer, but I actually imagine this service having a cleaner implementation with only the query builder interface.

mark-burnett commented on deckhand/objects/base.py:
```diff
@@ -0,0 +1,85 @@
+
+import datetime
+
+from oslo_versionedobjects import base
```
It's unclear to me what benefit we're getting from using oslo versionedobjects.

sh8121att replied:
It is a known danger with OVO that for documents with multi-line data that are whitespace sensitive, OVO will break them. Because Deckhand won't know which Kinds of documents this is, it seems like OVO requires us to Base64 encode everything to ensure the document sent to Deckhand is the document it returns. This breaks all manner of querying.

Commit:
Helper for generating versioned object automatically from dictionary payload.

Commit:
Update database documents schema.

Commit:
Added oslo_context-based context for oslo_db compatibility.

Commit:
Refactor database sqlalchemy api/models.

Commit:
Added logic for establishing DB connection.

Commit:
Successfully creating document.

TODO: clean up dummy test code and move registration of DB tables
to app initialization.

TODO: update documentation

TODO: include unit tests

To create a document:

1) run tox -egenconfig
2) rename the configured file from deckhand.conf.sample to
   deckhand.conf and place it in, e.g., ~/deckhand
3) include in deckhand.conf:

   [database]
   connection = sqlite:///:memory:

   to create an in-memory test db

Commit:
Cleaned up some logic, added exception handling to document creation.

Commit:
Change application/yaml to application/x-yaml

Commit:
Add currently necessary oslo namespaces to oslo-config-generator conf file.

This commit adds the following namespaces to deckhand's
config-generator.conf file used by oslo-config-generator
to generate deckhand's conf file automatically:

* oslo.db
* oslo.db.concurrency
* oslo.log

This will automatically populate the generate conf file with
needed oslo config options.

Commit:
Remove oslo_versionedobjects

This commit also updates documentation for launching test app.

Commit:
Add unit tests for db documents api.

Commit:
Use safe_load_all instead of safe_load.

fmontei committed:
Merge branch 'master' into documents-api

mark-burnett reviewed (approved):
Looks like a good skeleton for moving forward :)
